### PR TITLE
Rev swerve

### PR DESCRIPTION
**Pull request by falOn-Dev**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/swerve/Drivebase.kt [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+                    DriverStation.Alliance.Red -> Rotation2d.fromDegrees(180.0)
+                    DriverStation.Alliance.Blue -> Rotation2d()
+                    else -> Rotation2d(0.0)
+                }
+                hasAppliedOffset = true
+            }
+        }
+    }
+
+    fun setVisionSTDDevs(xMeters: Double, yMeters: Double, rotRads: Double) {
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/swerve/Drivebase.kt]: Measures should be used here

File: src/main/java/frc/robot/subsystems/swerve/Drivebase.kt [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        )
+
+
+        val brConfig: ModuleIO.ModuleConstants = ModuleIO.ModuleConstants(
+            10, 11, 12,
+            true, true, false,
+            Rotation2d.fromRotations(-0.247),
+            turnRatio, driveRatio,
+            wheelRadius
+        )
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/swerve/Drivebase.kt:331]: I'm not entirely sure, but these being differentiated by direction while in the rest of the subsystem being referenced by index doesn't sit right with me.

I would prefer if we just had a small inner class to wrap everything those arrays hold with named parts.

File: src/main/java/frc/robot/subsystems/swerve/gyro/GyroIO.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+interface GyroIO {
+    class GyroInputs : LoggableInputs {
+        var connected: Boolean = false
+        var yaw: Rotation2d = Rotation2d()
+        var yawVelocityDegreesPerSecond: Double = 0.0
+        var pitch: Rotation2d = Rotation2d()
+        var pitchVelocityDegreesPerSecond: Double = 0.0
+        var roll: Rotation2d = Rotation2d()
+        var rollVelocityDegreesPerSecond: Double = 0.0
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/swerve/gyro/GyroIO.kt]: Using Rotation2d is interesting; is it preferable to measures? Regardless, the velocities should definitely be measures

File: src/main/java/frc/robot/subsystems/swerve/gyro/GyroIO.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        override fun fromLog(table: LogTable?) {
+            table?.get("Connected")?.let { connected = it.boolean }
+            table?.get("YawDegrees", Rotation2d.struct, Rotation2d())?.let { yaw = it }
+            table?.get("YawVelocityDegreesPerSecond")?.let { yawVelocityDegreesPerSecond = it.double }
+            table?.get("PitchDegrees", Rotation2d.struct, Rotation2d())?.let { pitch = it }
+            table?.get("PitchVelocityDegreesPerSecond")?.let { pitchVelocityDegreesPerSecond = it.double }
+            table?.get("RollDegrees", Rotation2d.struct, Rotation2d())?.let { roll = it }
+            table?.get("RollVelocityDegreesPerSecond")?.let { rollVelocityDegreesPerSecond = it.double }
+        }
+    }
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/swerve/gyro/GyroIO.kt]: Use a non-null table parameter.

File: src/main/java/frc/robot/subsystems/swerve/gyro/GyroIO.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    fun updateInputs(inputs: GyroInputs) {}
+
+    fun setYaw(newYaw: Double) {}
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/swerve/gyro/GyroIO.kt]: Measure

File: src/main/java/frc/robot/subsystems/swerve/module/ModuleIO.kt [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        var driveStatorCurrent: Double = 0.0
+        var driveSupplyCurrent: Double = 0.0
+
+        var turnPosition: Rotation2d = Rotation2d()
+        var absoluteTurnPosition: Rotation2d = Rotation2d()
+        var turnVelocityRotationsPerSec: Double = 0.0
+        var turnSupplyVolts: Double = 0.0
+        var turnMotorVolts: Double = 0.0
+        var turnStatorCurrent: Double = 0.0
+        var turnSupplyCurrent: Double = 0.0
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/swerve/module/ModuleIO.kt:43]: I would have done a nested class, but I'm not gonna cry over it

File: src/main/java/frc/robot/subsystems/swerve/Drivebase.kt [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+        Constants.RobotConstants.Mode.SIM -> GyroIOSim(this::robotRelativeSpeeds)
+        Constants.RobotConstants.Mode.REPLAY -> object : GyroIO {}
+    }
+
+    /**
+     * Gyro inputs for reading values out of the gyro IO interface
+     */
+    val gyroInputs: GyroIO.GyroInputs = GyroIO.GyroInputs()
+
+    private val modules: Array<SwerveModule> = arrayOf(
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/swerve/Drivebase.kt:61]: I am extremely curious as to why this is an array.

File: src/main/java/frc/robot/subsystems/swerve/Drivebase.kt [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+        Translation2d(9.7859, 9.7859),
+        Translation2d(9.7859, -9.7859),
+        Translation2d(-9.7859, 9.7859),
+        Translation2d(-9.7859, -9.7859),
+    )
+
+    /**
+     * Array of [SwerveModulePosition] used for storing the distance traveled of each module, this is for odometry
+     */
+    private val modulePositions: Array<SwerveModulePosition> = Array(4) { SwerveModulePosition() }
```
Comment (Something-Inconspicuous) [inline src/main/java/frc/robot/subsystems/swerve/Drivebase.kt:91]: I think these three arrays can (and should) be consolidated into one array of a wrapper class for these modules, like a `SwerveModuleData` or something

**Review by Something-Inconspicuous — CHANGES_REQUESTED** [review]
I am not entirely sure when we should use `Measure<Angle>` vs `Rotation2d`.

The arrays of data in the swerve implementation seem to be able to be consolidated, so I would like that to happen.

**Review by mteam88 — APPROVED** [review]

**Pull request merged**